File: frame/src/main/java/com/dreampany/frame/data/model/Task.java
package com.dreampany.frame.data.model;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;

/**
 * Created by Hawladar Roman on 30/4/18.
 * Dreampany
 * dreampanymail@gmail.com
 */
public abstract class Task<T extends Parcelable> extends Base {
    
    protected String comment;
    protected T input;

    protected Task() {
    }

    protected Task(Parcel in) {
        super(in);
        comment = in.readString();
        if (in.readByte() == 0) {
            input = null;
        } else {
            Class<?> itemClazz = (Class<?>) in.readSerializable();
            input = in.readParcelable(itemClazz.getClassLoader());
        }
    }

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        dest.writeString(comment);
        if (input == null) {
            dest.writeByte((byte) 0);
        } else {
            dest.writeByte((byte) 1);
            Class itemClazz = input.getClass();
            dest.writeSerializable(itemClazz);
            dest.writeParcelable(input, flags);
        }
    }

    public void setComment(String comment) {
        this.comment = comment;
    }

    public void setInput(T input) {
        this.input = input;
    }

    public String getComment() {
        return comment;
    }

    public T getInput() {
        return input;
    }

}

File: frame/src/main/kotlin/com/dreampany/framework/data/model/Task.kt
package com.dreampany.framework.data.model

import android.os.Parcelable

/**
 * Created by roman on 2019-07-13
 * Copyright (c) 2019 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
abstract class Task<T : Parcelable>(open var input: T?, open var extra: String?) :
    BaseParcel() {

/*    var input: T? = null
    var extra: String? = null*/
/*    */

/*     constructor(`in`: Parcel) {
        extra = `in`.readString()
        if (`in`.readByte().toInt() == 0) {
            input = null
        } else {
            val itemClazz = `in`.readSerializable() as Class<*>
            input = `in`.readParcelable(itemClazz.classLoader)
        }
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(extra)
        if (input == null) {
            dest.writeByte(0.toByte())
        } else {
            dest.writeByte(1.toByte())
            val itemClazz = input!!.javaClass
            dest.writeSerializable(itemClazz)
            dest.writeParcelable(input, flags)
        }
    }

    override fun describeContents(): Int {
        return 0
    }*/
}